pil-analyzer/src/type_inference.rs

```diff
@@ -456,17 +457,21 @@ impl<'a> TypeChecker<'a> {
                 })
                 .map_err(|err| {
                     format!(
-                        "Expression is expected to evaluate to {} or ({})[]:\n  {expr}:\n{err}",
-                        expected_type.ty, expected_type.ty
+                        "Expression is expected to evaluate to {}:\n  {expr}:\n{err}",
+                        expected_type.ty
                     )
                 })
         } else {
-            self.expect_type(&expected_type.ty, expr)
+            self.expect_type(&expected_type.ty, expr, None)
         }
     }
 
     /// Process an expression and return the type of the expression.
-    fn infer_type_of_expression(&mut self, e: &mut Expression) -> Result<Type, String> {
+    fn infer_type_of_expression(
+        &mut self,
+        e: &mut Expression,
+        kind: Option<FunctionKind>,
```

Comment:
I think it would be better to store that inside TypeChecker instead of dragging it around all the time. We also store local variable types, and this is kind of similar.